src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -708,6 +727,31 @@ public IEETypeNode NecessaryTypeSymbol(TypeDesc type)
             return _typeSymbols.GetOrCreateValue(type);
         }
 
+        private sealed class MetadataTypeSymbolHashtable : TypeSymbolHashtable
+        {
+            public MetadataTypeSymbolHashtable(NodeFactory factory) : base(factory) { }
+            protected override IEETypeNode CreateValueFromKey(TypeDesc key) => _factory.CreateMetadataTypeNode(key);
+        }
+
+        private MetadataTypeSymbolHashtable _metadataTypeSymbols;
+
+        public IEETypeNode MetadataTypeSymbol(TypeDesc type)
+        {
+            if (_compilationModuleGroup.ShouldReferenceThroughImportTable(type))
```

Comment:
Why do we use `ShouldReferenceThroughImportTable` here and `ContainsType` for the constructed type node?